usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -4414,6 +4415,48 @@ int64_t arc_pages_pp_reserve = 64;
  */
 int64_t arc_swapfs_reserve = 64;
 
+static volatile uint64_t arc_virt_machine_reserved;
+
+/*
+ * XXX: A possible concern is that we allow arc_virt_machine_reserved to
+ * get so large that we cause the arc to perform a lot of additional
+ * work to keep the arc extremely small. We may want to set limits to
+ * the size of arc_virt_machine_reserved and disallow reservations
+ * beyond that limit.
+ */
+int
+arc_virt_machine_reserve(size_t pages)
+{
+	uint64_t newv;
+
+	newv = atomic_add_64_nv(&arc_virt_machine_reserved, pages);
+
+	/*
+	 * Since arc_virt_machine_reserved effectively lowers arc_c_max
+	 * as needed for vmm memory, if this request would put the arc
+	 * under arc_c_min, we reject it.  arc_c_min should be a value that
+	 * ensures reasonable performance for non-VMM stuff, as well as keep
+	 * us from dipping below lotsfree, which could trigger the pager
+	 * (and send the system toa grinding halt while it pages).
+	 *
+	 * XXX: This is a bit hacky and might be better done w/ a mutex
+	 * instead of atomic ops.
+	 */
+	if (newv + arc_c_min > arc_c_max) {
+		atomic_add_64(&arc_virt_machine_reserved, -(int64_t)pages);
```

Comment:
I am not entirely sure how atomics work in regards to barriers of other threads seeing values change.  Is it possible that the atomic add above has already caused the arc to start flushing things before we get to this safety check?

Reply:
The use of atomics is to guard against concurrent threads using arc_virt_machine_reserve from stepping on each others' view of arc_virt_machine_reserved.

The read of arc_virt_machine_reserved in arc_available_memory() isn't guarded by anything. It means a thread there sees whatever it does at the time.  AIUI that function is either called by kmem (which in the worst case is either too aggressive or not aggressive enough, the latter is what we have today) or by arc_reap_zthr, which is woken up AFTER we do maniuplations here.